backend/src/main/java/net/es/oscars/web/rest/VirtInterfaceController.java

```diff
@@ -0,0 +1,283 @@
+
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.google.common.net.InetAddresses;
+
+import lombok.extern.slf4j.Slf4j;
+import net.es.oscars.resv.ent.Connection;
+import net.es.oscars.resv.ent.VlanFixture;
+import net.es.oscars.resv.svc.ConnService;
+import net.es.oscars.sb.nso.db.NsoVirtInterfaceDAO;
+import net.es.oscars.sb.nso.ent.NsoVirtInterface;
+import net.es.oscars.web.beans.VirtIpInterfaceListResponse;
+import net.es.oscars.web.beans.VirtIpInterfaceRequest;
+
+@Slf4j
+@RestController
+public class VirtInterfaceController {
+
+    private final NsoVirtInterfaceDAO virtInterfaceDAO;
+    private final ConnService connSvc;
+
+
+    public VirtInterfaceController(NsoVirtInterfaceDAO virtInterfaceDAO,
+                                    ConnService connSvc) {
+        this.virtInterfaceDAO = virtInterfaceDAO;
+        this.connSvc = connSvc;
+    }
+
+
+    @RequestMapping(value = "/api/interface/add", method = RequestMethod.POST)
+    @ResponseBody
+    @Transactional
+    public void addVirtIpInterface(@RequestBody VirtIpInterfaceRequest request) {
+        log.debug("request: " + request.toString());
+
+        verifyData(request);
+
+        String connectionId = request.getConnectionId();
+        String device = request.getDevice();
+        String ipAndSubnet = request.getIpAndSubnet();
+
+        List<NsoVirtInterface> interfaces = virtInterfaceDAO.findByConnectionId(connectionId);
+
+        // if no instances exist: create first one for circuit and add virt ip
+        if (interfaces == null || interfaces.isEmpty()) {
+            log.info("No existing instances for connection id" + connectionId + " found");
+            NsoVirtInterface newEntry = new NsoVirtInterface();
+            newEntry.setDevice(device);
+            newEntry.setConnectionId(connectionId);
+            newEntry.getIpAddresses().add(ipAndSubnet);
+            virtInterfaceDAO.save(newEntry);
+            log.info("Virt IP " + ipAndSubnet + " added to " + device + " for connection id " + connectionId);
+            return;
+        }
+
+        // if entries exist:
+        // 1. check all entries if ip/net already exists
+        // 2. if yes -> error out | if no -> create / add entry
+        boolean ipExists = false;
+        HashMap<String, NsoVirtInterface> cache = new HashMap<String, NsoVirtInterface>();
+        for (NsoVirtInterface ifce : interfaces) {
+            cache.put(ifce.getDevice(), ifce);
+            for (String existingIp : ifce.getIpAddresses()) {
+                if (existingIp.equals(ipAndSubnet)) {
```

Comment:
yeah, so the question is if we add an additional check for the IP "only" part of this IP/mask combo?

Reply:
I don't see it as mandatory to do that. but hey, bonus points if you do! should be easy enough; I recommend using  https://github.com/seancfoley/IPAddress to do this validation

Reply:
I tried to use this lib but it has some wildcard stuff etc. and ended up to introduce new "bugs" - e.g. you could submit an IP like 10.0.0/24 and it was valid (I'm not saying that it's not ... ) IDK

I ended up just introducing a return type (IP string) to the check CIDR method I had